torch/_dynamo/guards.py

```diff
@@ -100,6 +100,29 @@ def strip_getattr_getitem(name):
     return re.split(r"[.\[]", name)[0]
 
 
+# Don't put anything on here you don't want laundered to the next frame
+@dataclasses.dataclass
+class CodePart:
+    source: Optional[
+        List[Source]
+    ]  # Note: List of sources for tensor checks and shape_env
+    code: str
+    origin: str
+
+    # bound at guard failure time
+    scope = None
+
+    # tensor check only
+    check_tensor_verbose = None
+    tensor_check_names: Optional[str] = None
+
+    def __hash__(self):
+        return hash(self.code)
```

Comment:
Does this imply that for any code, there is only ever one CodePart? Can the non code fields vary for the same code?

Reply:
For a given code, all the other parts should be identical. So, yes. I do not know a case where this is not sound.

Reply:
Seems like a good invariant to write as a comment